vscode/vscode/vscode.ml

```diff
@@ -2521,6 +2652,271 @@ module TreeView = struct
+
+  val set_webview : t -> WebView.t -> unit
+
+  val dispose : t -> Js.Any.t
+
+  val reveal :
+    t -> ?preserveFocus:bool -> ?viewColumn:ViewColumn.t -> unit -> unit
+
+  val create :
+       onDidChangeViewState:WebviewPanelOnDidChangeViewStateEvent.t Event.t
+    -> onDidDispose:unit Event.t
+    -> active:bool
+    -> options:WebviewPanelOptions.t
+    -> title:string
+    -> viewColumn:ViewColumn.t
+    -> viewType:string
+    -> visible:bool
+    -> webview:WebView.t
+    -> dispose:Js.Any.t
+    -> reveal:(?preserveFocus:bool -> ?viewColumn:ViewColumn.t -> unit -> unit)
+    -> t
+end = struct
+  include Interface.Make ()
+
+  module LightDarkIcon = struct
+    type t =
+      { light : ([ `Uri of Uri.t ][@js.union])
+      ; dark : ([ `Uri of Uri.t ][@js.union])
+      }
+    [@@js]
+
+    let t_of_js js_val =
+      let light_js = Ojs.get_prop_ascii js_val "light" in
+      let dark_js = Ojs.get_prop_ascii js_val "dark" in
+      let light = `Uri ([%js.to: Uri.t] light_js) in
+      let dark = `Uri ([%js.to: Uri.t] dark_js) in
+      { light; dark }
+  end
+
+  type iconPath =
+    ([ `Uri of Uri.t
+     | `LightDark of LightDarkIcon.t
+     ]
+    [@js.union])
+  [@@js]
+
+  module OnDidChangeViewState =
+    Event.Make (WebviewPanelOnDidChangeViewStateEvent)
+  module OnDidDispose = Event.Make (Js.Unit)
+
+  include
+    [%js:
+    val onDidChangeViewState : t -> OnDidChangeViewState.t [@@js.get]
+
+    val onDidDispose : t -> OnDidDispose.t [@@js.get]
+
+    (*let iconPath_of_js js_val = if Ojs.has_property js_val "path" then `Uri
+      ([%js.to: Uri.t] js_val) else if Ojs.has_property js_val "light" then
+      `LightDark ([%js.to: LightDarkIcon.t] js_val) else assert false*)
+
+    val active : t -> bool [@@js.get]
+
+    val options : t -> WebviewPanelOptions.t [@@js.get]
+
+    val title : t -> string [@@js.get]
+
+    val viewColumn : t -> ViewColumn.t [@@js.get]
+
+    val viewType : t -> string [@@js.get]
+
+    val visible : t -> bool [@@js.get]
+
+    val webview : t -> WebView.t [@@js.get]
+
+    val set_webview : t -> WebView.t -> unit [@@js.set]
+
+    val dispose : t -> Js.Any.t [@@js.call]
+
+    val reveal :
+      t -> ?preserveFocus:bool -> ?viewColumn:ViewColumn.t -> unit -> unit
+      [@@js.call]
+
+    val create :
+         onDidChangeViewState:OnDidChangeViewState.t
+      -> onDidDispose:OnDidDispose.t
+      -> active:bool
+      -> options:WebviewPanelOptions.t
+      -> title:string
+      -> viewColumn:ViewColumn.t
+      -> viewType:string
+      -> visible:bool
+      -> webview:WebView.t
+      -> dispose:Js.Any.t
+      -> reveal:
+           (?preserveFocus:bool -> ?viewColumn:ViewColumn.t -> unit -> unit)
+      -> t
+      [@@js.builder]]
+end
+
+and WebviewPanelOnDidChangeViewStateEvent : sig
+  include Js.T
+
+  val webviewPanel : t -> WebviewPanel.t
+end = struct
+  include Interface.Make ()
+
+  include [%js: val webviewPanel : t -> WebviewPanel.t [@@js.get]]
+end
```

Comment:
If you place this module inside the `WebviewPanel` module, it would remove the need for using recursive modules.

Reply:
Is it ok to have `WebviewPanelOnDidChangeViewStateEvent .webviewPanel` have signature `: t -> t`? 
https://github.com/ocamllabs/vscode-ocaml-platform/commit/ef4ee16416ee5e3430cfcb0e8015b853895750a4#diff-f66b1e1ddae54b2c611b3c724af7f1c92c192c722a2bb03ea746cab601b9553bR2739,
Or should it be something else?

Reply:
Thanks, 
I also had to add `[@@js]` to type alias 
 https://github.com/ocamllabs/vscode-ocaml-platform/pull/666/commits/5844d1b6545e2a15cdef8839aa9540638337036f#diff-f66b1e1ddae54b2c611b3c724af7f1c92c192c722a2bb03ea746cab601b9553bR2740 
Is this correct? 
Or should i add `let webviewPanel_of_js = t_of_js` right after it?

Reply:
Oops, forget about that.
The `[@@js]` is fine.